observatory-dags/observatory/dags/telescopes/jstor.py

```diff
@@ -0,0 +1,451 @@
+        :return: None.
+        """
+
+        # Upload each downloaded release
+        for release in releases:
+            upload_files_from_list(release.download_files, release.download_bucket)
+
+    def transform(self, releases: List[JstorRelease], **kwargs):
+        """ Task to transform the Jstor releases for a given month.
+
+        :param releases: a list of Jstor releases.
+        :return: None.
+        """
+
+        # Transform each release
+        for release in releases:
+            release.transform()
+
+
+@retry(stop=stop_after_attempt(JstorTelescope.MAX_ATTEMPTS),
+       reraise=True,
+       wait=wait_fixed(JstorTelescope.FIXED_WAIT) + wait_exponential(multiplier=JstorTelescope.MULTIPLIER,
+                                                                     exp_base=JstorTelescope.EXP_BASE,
+                                                                     max=JstorTelescope.MAX_WAIT_TIME),
+       )
+def download_report(url: str, download_path: str) -> bool:
+    """ Download report from url to a file.
+
+    :param url: Download url
+    :param download_path: Path to download data to
+    :return: Whether download was successful
+    """
+    logging.info(
+        f'Downloading report from url, attempt: {download_report.retry.statistics["attempt_number"]}, idle for:'
+        f'{download_report.retry.statistics["idle_for"]}')
+    response = requests.get(url)
+    if response.status_code != 200:
+        raise AirflowException(f'Could not download content from report url, reason: {response.reason}, '
+                               f'status_code: {response.status_code}')
+
+    content = response.content.decode('utf-8')
+    with open(download_path, 'w') as f:
+        f.write(content)
+    return True
+
+
+@retry(stop=stop_after_attempt(JstorTelescope.MAX_ATTEMPTS),
+       reraise=True,
+       wait=wait_fixed(JstorTelescope.FIXED_WAIT) + wait_exponential(multiplier=JstorTelescope.MULTIPLIER,
+                                                                     exp_base=JstorTelescope.EXP_BASE,
+                                                                     max=JstorTelescope.MAX_WAIT_TIME),
+       )
+def get_header_info(url: str) -> [str, str]:
+    """ Get header info from url and parse for filename and extension of file.
+
+    :param url: Download url
+    :return: Filename and file extension
+    """
+    logging.info(f'Getting HEAD of report download url, attempt: {get_header_info.retry.statistics["attempt_number"]}, '
+                 f'idle for: {get_header_info.retry.statistics["idle_for"]}')
+    response = requests.head(url, allow_redirects=True)
+    if response.status_code != 200:
+        raise AirflowException(f'Could not get HEAD of report download url, reason: {response.reason}, '
+                               f'status_code: {response.status_code}')
+    filename, extension = response.headers['Content-Disposition'].split('=')[1].split('.')
+    return filename, extension
```

Comment:
Nice, these look good.

I'm sorry I misunderstood something in the previous review and said that the create_headers function was not needed when calling requests.head, which was wrong. So it would be good to add the following function back and use it to set the headers in requests.get and requests.head in download_report and get_header_info respectively:

```python
def create_headers(url: str) -> dict:
    """ Create a headers dict that can be used to make a request
    :param url: the download url
    :return: headers dictionary
    """
    referer = f"https://www.google.com/url?q={url}" \
              f"&amp;source=gmail&amp;ust={int(time.time())}000&amp;usg=AFQjCNFtACM-4Zqs3yA1AXl4GyEbfvCqwQ"
    headers = {
        'Referer': referer,
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) '
                      'Chrome/89.0.4389.90 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,'
                  'application/signed-exchange;v=b3;q=0.9'

    }
    return headers
```

Reply:
Ah okay, did you find out that it failed without the headers?

Reply:
Yeah it failed without them